tooling/miniutil/src/build/terminator.rs

```diff
@@ -1,5 +1,243 @@
+
+    pub fn atomic_load(&mut self, dest: PlaceExpr, ptr: ValueExpr) {
+        self.terminator = Some(atomic_load(dest, ptr, self.next_block));
+        self.finish_block();
+        self.prepare_next_block();
+    }
+
+    pub fn atomic_fetch(
+        &mut self,
+        binop: FetchBinOp,
+        dest: PlaceExpr,
+        ptr: ValueExpr,
+        other: ValueExpr,
+    ) {
+        self.terminator = Some(atomic_fetch(binop, dest, ptr, other, self.next_block));
+        self.finish_block();
+        self.prepare_next_block();
+    }
+
+    pub fn compare_exchange(
+        &mut self,
+        dest: PlaceExpr,
+        ptr: ValueExpr,
+        current: ValueExpr,
+        next_val: ValueExpr,
+    ) {
+        self.terminator = Some(compare_exchange(dest, ptr, current, next_val, self.next_block));
+        self.finish_block();
+        self.prepare_next_block();
+    }
+
+    pub fn expose_provenance(&mut self, dest: PlaceExpr, ptr: ValueExpr) {
+        self.terminator = Some(expose_provenance(dest, ptr, self.next_block));
+        self.finish_block();
+        self.prepare_next_block();
+    }
+
+    pub fn with_exposed_provenance(&mut self, dest: PlaceExpr, addr: ValueExpr) {
+        self.terminator = Some(with_exposed_provenance(dest, addr, self.next_block));
+        self.finish_block();
+        self.prepare_next_block();
+    }
+
+    pub fn lock_create(&mut self, ret: PlaceExpr) {
+        self.terminator = Some(lock_create(ret, self.next_block));
+        self.finish_block();
+        self.prepare_next_block();
+    }
+
+    pub fn lock_acquire(&mut self, lock_id: ValueExpr) {
+        self.terminator = Some(lock_acquire(lock_id, self.next_block));
+        self.finish_block();
+        self.prepare_next_block();
+    }
+
+    pub fn lock_release(&mut self, lock_id: ValueExpr) {
+        self.terminator = Some(lock_release(lock_id, self.next_block));
+        self.finish_block();
+        self.prepare_next_block();
+    }
+
+    // terminators with 2 or more following blocks
+    pub fn if_<F, G>(&mut self, condition: ValueExpr, then_branch: F, else_branch: G)
+    where
+        F: Fn(&mut Self),
+        G: Fn(&mut Self),
+    {
+        self.switch_int(bool_to_int::<u8>(condition), &[(1, &then_branch)], else_branch);
+    }
+
+    pub fn switch_int<T, G>(
+        &mut self,
+        value: ValueExpr,
+        cases: &[(T, &dyn Fn(&mut Self))],
+        fallback: G,
+    ) where
+        T: Clone + Into<Int>,
+        G: Fn(&mut Self),
+    {
+        let cases: Vec<(&(T, &dyn Fn(&mut Self)), u32)> =
+            cases.iter().zip(self.next_block..).collect();
+        let cases_nr: Vec<(T, u32)> =
+            cases.iter().map(|((case, _), successor)| (case.clone(), successor.clone())).collect();
+
+        self.next_block += cases.len() as u32;
+        let fallback_nr = self.next_block;
+        // reserve next block nr -> if not used, function ends anyway
+        let next = self.next_block + 1;
+        self.next_block += 2;
```

Comment:
This seems like pretty fragile reasoning for block number freshness, the `declare_block` API should help simplify that I hope.